File: dart/collatz-conjecture/lib/collatz_conjecture.dart
class CollatzConjecture {
  int steps(final int number) {
    if (number < 1) {
      throw ArgumentError('Only positive integers are allowed');
    }

    int cnt = 0;
    int n = number;

    while (n != 1) {
      if (n.isEven) {
        n ~/= 2;
      } else {
        n = 3 * n + 1;
      }
      cnt++;
    }

    return cnt;
  }
}

File: cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture {

auto steps(int n) -> int {
  if (n <= 0) {
    throw std::domain_error("Input must be a positive integer!");
  }

  int cnt = 0;
  while (n != 1) {
    if (n % 2 == 0) {
      n /= 2;
    } else {
      n = 3 * n + 1;
    }
    ++cnt;
  }
  return cnt;
}

} // namespace collatz_conjecture
